File: tree.c
#include "tree.h"

struct Tree {
    TreeNode *root;
};

Tree *create_tree() {
    Tree *tree = (Tree*) malloc(sizeof(Tree));
    if(tree == NULL) return NULL;
    tree->root = NULL;
    return tree;
}

static TreeNode *insert_node(TreeNode *root, TreeNode *parent, int val) {
    if(root == NULL) {
        root = (TreeNode*) malloc(sizeof(TreeNode));
        if(root == NULL) return NULL;
        root->parent = parent;
        root->left = root->right = NULL;
        root->val = val;
    } else {
        if(val < root->val) {
            root->left = insert_node(root->left, root, val);
        } else if(val > root->val) {
            root->right = insert_node(root->right, root, val);
        }
    }
    return root;
}

void insert_tree(Tree *tree, int val) {
    tree->root = insert_node(tree->root, NULL, val);
}

static void print_nodes(TreeNode *root, FILE *out) {
    if(root == NULL) return;

    print_nodes(root->left, out);
    fprintf(out, "%d ", root->val);
    print_nodes(root->right, out);
}

void print_tree(Tree *tree, FILE *out) {
    print_nodes(tree->root, out);
}

static void free_nodes(TreeNode *root) {
    if (root == NULL)
        return;

    free_nodes(root->left);
    free_nodes(root->right);
    free(root);
}

void clear_tree(Tree *tree) {
    free_nodes(tree->root);
    tree->root = NULL;
}

void delete_tree(Tree *tree) {
    clear_tree(tree);
    free(tree);
}

File: Py/datastructures/tree.py
class TreeNode:
    def __init__(self, data, parent=None, left=None, right=None):
        self.data = data
        self.parent = parent
        self.left = left
        self.right = right
    def __str__(self):
        return str(self.data)

class Tree:
    def __init__(self):
        self.root = None
    def node_insert(self, root, parent, data):
        if root == None:
            return TreeNode(data, parent)
        else:
            if data < root.data:
                root.left = self.node_insert(root.left, root, data)
            elif data > root.data:
                root.right = self.node_insert(root.right, root, data)
            return root
    def insert(self, data):
        self.root = self.node_insert(self.root, None, data)
    def node_search(self, root, data):
        if root == None:
            return False
        if root.data == data:
            return True
        elif root.data > data:
            return self.node_search(root.left)
        else:
            return self.node_search(root.right)
    def search(self, data):
        return self.node_search(self.root, data)
    def node_print(self, root):
        if root == None:
            return ''
        s = ''
        tmp = self.node_print(root.left)
        if tmp != '':
            s += tmp + ' '
        s += str(root.data)
        tmp = self.node_print(root.right)
        if tmp != '':
            s += ' ' + tmp
        return s
    def __str__(self):
        return self.node_print(self.root)
    def clear(self):
        self.root = None

        
